### ISSUE TITLE
No inference for TypeVar in Callable (aka unable to annotate decorator factory)

### ISSUE THREAD

**Issue by Drino** [microsoft/pyright#3788]
**Describe the bug**
I'm trying to write type hints for a function that returns a decorator. 

Unfortunately, I've not managed to write a type for a decorator. I suppose that it should be as following:
```python
TCallable = TypeVar('TCallable', bound=Callable[..., Any])
TDecorator: TypeAlias = Callable[[TCallable], TCallable]
```

But anything decorated with a `TDecorator` object is inferred as `Unknown`. 

**To Reproduce**
Run pylance on this snippet:
```python
from typing import TypeAlias, TypeVar, Any, Callable
from typing_extensions import reveal_type

def fn() -> int: return 42


TCallable = TypeVar('TCallable', bound=Callable[..., Any])

TDecorator: TypeAlias = Callable[[TCallable], TCallable]

noop: TDecorator = lambda x: x
just_fn = noop(fn)
reveal_type(just_fn)  # Type of "just_fn" is "Unknown"
```

**Expected behavior**
```python
reveal_type(just_fn)  # Type of "just_fn" is "() -> int"
```

**VS Code extension or command-line**
This reproduces in both the vs code extension and the cli (1.1.264 and earlier versions). 

**Additional context**
Actually, I'm trying to get correct hints for this code:
```python
from typing import Generic, TypeVar, Any, Callable

TContainedCallable = TypeVar('TContainedCallable', bound=Callable[..., Any])


class Container(Generic[TContainedCallable]):
    def __init__(self, callable: TContainedCallable):
        self.callable = callable


TCallable = TypeVar('TCallable', bound=Callable[..., Any])

TContainerMaker = Callable[[TCallable], Container[TCallable]]

maker: TContainerMaker = lambda c: Container(c)

TContainerUnpacker = Callable[[Container[TCallable]], TCallable]

unpacker: TContainerUnpacker = lambda c: c.callable


def answer() -> int:
    return 42
    

answer_container = maker(answer)
reveal_type(answer_container)
unpacked_answer = unpacker(answer_container)
reveal_type(unpacked_answer)
```

**Comment by erictraut** [comment]
This is working as designed. You are using generic type aliases and type variables incorrectly here. If you enable `reportMissingTypeArguments`, pyright will flag the error.

The problem is that you're defining a generic type alias called `TDecorator` that has a single type parameter. That means whenever you use this type alias, you must supply a type argument for it. If you don't supply a type argument, it is assumed to be an implicit `Any` (what pyright refers to as `Unknown`).

When you declare the type of the variable `noop`, you could do the following:
```python
noop: TDecorator[Callable[..., int]] = lambda x: x
```

The other thing to consider is that type variables have meaning only when they are scoped to a class, function, or type alias. A type variable cannot be scoped to a variable. A variable's type annotation can refer to a type variable, but that type variable needs to be scoped to a function or class that contains the variable.

In your example above, I recommend moving from a lambda to a full function declaration, where you can supply proper type annotations for the parameters. The function will supply the necessary scope for the type variable.

```python
def maker(c: TContainedCallable) -> Container[TContainedCallable]:
    return Container(c)
```

**Comment by Drino** [comment]
Thank you very much for clarification! 

> I recommend moving from a lambda to a full function declaration, where you can supply proper type annotations for the parameters. The function will supply the necessary scope for the type variable.

Alas, this approach doesn't work for me, as I have decorator factories, like in click:
```
@click.command()
@click.option('--count', default=1, help='Number of greetings.')
@click.option('--name', prompt='Your name',
              help='The person to greet.')
def hello(count, name):
    ...
```

I've managed to overcome this using protocols:
```python
from typing import Generic, TypeVar, Any, Callable, Protocol

TContainedCallable = TypeVar('TContainedCallable', bound=Callable[..., Any])


class Container(Generic[TContainedCallable]):
    def __init__(self, callable: TContainedCallable):
        self.callable = callable


TCallable = TypeVar('TCallable', bound=Callable[..., Any])


class TContainerMaker(Protocol):
    def __call__(self, _: TCallable, /) -> Container[TCallable]:
        ...


class TContainerUnpacker(Protocol):
    def __call__(self, _: Container[TCallable], /) -> TCallable:
        ...

def maker_factory() -> TContainerMaker:
    return lambda c: Container(c)


def unpacker_factory() -> TContainerUnpacker:
    return lambda c: c.callable

unpacker: TContainerUnpacker = unpacker_factory()
maker: TContainerMaker = maker_factory()


def answer() -> int:
    return 42


answer_container = maker(answer)
reveal_type(answer_container)  # Type of "answer_container" is "Container[() -> int]"
unpacked_answer = unpacker(answer_container)
reveal_type(unpacked_answer)  # Type of "unpacked_answer" is "() -> int"
```

But for a strange reason, if I inline lambdas, i.e.:
```python
maker: TContainerMaker = lambda c: Container(c)
unpacker: TContainerUnpacker = lambda c: c.callable
```

I get:
```python
answer_container = maker(answer)  
# Argument of type "() -> int" cannot be assigned to parameter "c" of type "TCallable@__call__"
# Type "() -> int" cannot be assigned to type "TCallable@__call__"
reveal_type(answer_container)
unpacked_answer = unpacker(answer_container)
reveal_type(unpacked_answer)  
# Argument of type "Container[TCallable@__call__]" cannot be assigned to parameter "c" of type "Container[TCallable@__call__]"
#  TypeVar "TContainedCallable@Container" is invariant
#    Type "TCallable@__call__" cannot be assigned to type "TCallable@__call__"
```

**Comment by erictraut** [comment]
I'm not able to repro the errors you're seeing. I've copied the two blocks of code from your previous post (the code snippets that are both above and below "But for a strange reason...") into a new ".py" file, and I don't see any errors. Is there something else I need to do to repro?

**Comment by Drino** [comment]
Sorry for the inconvenience. To reproduce the issue it is required to replace definitions of `maker` and `unpacker`. 

The whole snippet is like this:
```python
from typing import Generic, TypeVar, Any, Callable, Protocol

TContainedCallable = TypeVar('TContainedCallable', bound=Callable[..., Any])


class Container(Generic[TContainedCallable]):
    def __init__(self, callable: TContainedCallable):
        self.callable = callable


TCallable = TypeVar('TCallable', bound=Callable[..., Any])


class TContainerMaker(Protocol):
    def __call__(self, _: TCallable, /) -> Container[TCallable]:
        ...


class TContainerUnpacker(Protocol):
    def __call__(self, _: Container[TCallable], /) -> TCallable:
        ...

maker: TContainerMaker = lambda c: Container(c)
unpacker: TContainerUnpacker = lambda c: c.callable


def answer() -> int:
    return 42


answer_container = maker(answer)
reveal_type(answer_container)
# Argument of type "() -> int" cannot be assigned to parameter "c" of type "TCallable@__call__"
# Type "() -> int" cannot be assigned to type "TCallable@__call__"
unpacked_answer = unpacker(answer_container)
reveal_type(unpacked_answer)
# Argument of type "Container[TCallable@__call__]" cannot be assigned to parameter "c" of type "Container[TCallable@__call__]"
#  TypeVar "TContainedCallable@Container" is invariant
#    Type "TCallable@__call__" cannot be assigned to type "TCallable@__call__"
```

**Comment by erictraut** [comment]
There are two way that type variables can be replaced by a concrete type:

1. Explicitly through specialization by providing a list of type arguments in a subscript expression, for example `list[int]`.
2. Through a call expression where the type checker "solves" the type variable by identifying a type that satisfies all of the constraints. Type variables must be solved within the scope they are defined. If the type variable is scoped to a function, it is solved when you call that function. If a type variable is scoped to a class, it is solved when you call the constructor for that class.

In your example above, you have a function-scoped type variable in the `TContainerMaker.__call__` method, but there is no call expression that invokes this method. That means the type variable `TCallable@__call__` is never "solved".

I think what you perhaps meant to do here is to define a class-scoped type variable for the protocol classes `TContainerMaker` and `TContainerUnpacker`, like this:

```python
class TContainerMaker(Protocol[TCallable]):
    def __call__(self, _: TCallable, /) -> Container[TCallable]:
        ...

class TContainerUnpacker(Protocol[TCallable]):
    def __call__(self, _: Container[TCallable], /) -> TCallable:
        ...

maker: TContainerMaker[Callable[..., Any]] = lambda c: Container(c)
unpacker: TContainerUnpacker[Callable[..., Any]] = lambda c: c.callable
```

Is that what you intended?

**Comment by Drino** [comment]
Thank you for the fast answer!

> In your example above, you have a function-scoped type variable in the TContainerMaker.__call__ method, but there is no call expression that invokes this method.

```
# I hint TContainerMaker as type of maker here:
maker: TContainerMaker = lambda c: Container(c)
...
# And then I call it here:
answer_container = maker(answer)
reveal_type(answer_container)
```

The full snippet that demonstrates this specific behavior for lambdas:
```
from typing import Generic, TypeVar, Any, Callable, Protocol
from typing_extensions import reveal_type

TContainedCallable = TypeVar('TContainedCallable', bound=Callable[..., Any])


class Container(Generic[TContainedCallable]):
    def __init__(self, callable: TContainedCallable):
        self.callable = callable


TCallable = TypeVar('TCallable', bound=Callable[..., Any])


# Let's write regular decorator functions!
def maker_fn(c: TCallable) -> Container[TCallable]:
    return Container(c)


def unpacker_fn(c: Container[TCallable]) -> TCallable:
    return c


@unpacker_fn
@maker_fn
def using_fn() -> int:
    return 42


# Let's write protocol for decorators and factories that provide them!

class TContainerMaker(Protocol):
    def __call__(self, _: TCallable, /) -> Container[TCallable]:
        ...


class TContainerUnpacker(Protocol):
    def __call__(self, _: Container[TCallable], /) -> TCallable:
        ...


def maker_factory() -> TContainerMaker:
    return lambda c: Container(c)


def unpacker_factory() -> TContainerUnpacker:
    return lambda c: c.callable


@unpacker_factory()
@maker_factory()
def using_factories() -> int:
    return 42


# Let's just assign lambdas to variables.


maker_lambda: TContainerMaker = lambda c: Container(c)
unpacker_lambda: TContainerUnpacker = lambda c: c.callable


@unpacker_lambda
@maker_lambda
def using_lambdas() -> int:
    return 42


# Let's assign functions to variables.

def _maker_fn(c):
    return Container(c)


def _unpacker_fn(c):
    return c.callable

maker_fn_assign: TContainerMaker = _maker_fn
unpacker_fn_assign: TContainerUnpacker = _unpacker_fn

@unpacker_fn_assign
@maker_fn_assign
def using_fn_assign() -> int:
    return 42



reveal_type(using_fn)  # Type of "using_fn" is "() -> int"
reveal_type(using_factories)  # Type of "using_factories" is "() -> int"
# Next line seems to be broken
reveal_type(using_lambdas)  # Type of "using_lambdas" is "TCallable@__call__"
reveal_type(using_fn_assign)  # Type of "using_fn_assign" is "() -> int"
```

If I enable `"python.analysis.typeCheckingMode": "basic" ` in vscode using Pylance v2022.8.11, I'll get also the errors from my previous post on `@unpacker/maker_lambda` applications. This looks like some sort of a bug is involved.

> Is that what you intended?

Thank you for the clarification here, but I'm not sure. First of all, if I rewrite snippet from above using class-scope typevar and change annotations for `maker_lambda/unpacker_lambda` like suggested:
```
maker_lambda: TContainerMaker[Callable[..., Any]] = lambda c: Container(c)
unpacker_lambda: TContainerUnpacker[Callable[..., Any]] = lambda c: c.callable
```
I will get 
```
reveal_type(using_lambdas)  # Type of "using_lambdas" is "(...) -> Any"
```
which is expected, but not the desired behavior. 

I will actually get the desired behavior if I 

There is also a conceptual problem here. I may be wrong, but I've thought that the correct protocol for a regular decorator (the kind that does not affect the function signature) is 
```
TCallable = TypeVar('TCallable', bound=Callable[..., Any])

class TDecorator(Protocol):
    def __call__(self, _: TCallable, /) -> TCallable:
        ...
```

It can be applied to (called on) any function and return a compatible function/wrapper. The type of `TCallable` will be identified as the type of a decorated function (at least I expect no other restrictions on `TCallable`).

But if `TCallable` is class-scoped, I'd expect it to be the same for all calls of an instance of that class. Thus, return type may be different from argument type, but they both should be compatible with some instance-level `TCallable` value, like here:
```
class RandomFunctionPicker(Generic[TCallable]):
    def __init__(self):
        self.functions: list[TCallable] = []

    def __call__(self, _: TCallable, /) -> TCallable:
        self.functions.append(_)
        return random.choice(self.functions)
```

Though, if we don't provide the class variable, it looks like pyright treat both Unknown class-scoped variable and method-scoped variable the same (or at least very similar) way, e.g.:

```
picker = PickRandomFunction()


@picker
def just_fn(_) -> int:
    return 42


@picker
def my_fn(_, x=42) -> int:
    return x


@picker
def some_fn(x=42) -> int:
    return x


reveal_type(just_fn)  # Type of "just_fn" is "(_: Unknown) -> int"
reveal_type(my_fn)  # Type of "my_fn" is "(_: Unknown, x: int = 42) -> int"
reveal_type(some_fn)  # Type of "some_fn" is "(x: int = 42) -> int"
reveal_type(picker)  # Type of "picker" is "PickRandomFunction[Unknown]"
```
But the correct type of `just_fn/my_fn/some_fn` is `Unknown` (or `Callable[[Any], int]` but this can't be inferred). This is not a bug report, just an observation.

**Comment by erictraut** [comment]
I'm not sure I follow everything in your last post because some of the code snippets are incomplete, but when looking at the snippets that are complete, I don't see any evidence of a bug. Pyright is working as intended.

Consider the following code snippet:

```python
T = TypeVar("T")
v: list[T] = [1, 2, 3]
```

This code has a problem in that the annotation for variable `v` is using an unscoped and unsolved type variable. I think you'd agree this is a problem.

Now, let's consider a similar code snippet:

```python
T = TypeVar("T")
v: Callable[[T], T] = lambda x: x
```

Here again, `v` is annotated with a type that has an unsolved type variable `T`. Do you agree this is a problem?

What you're doing in your code is effectively the same as this. You're replacing the `Callable` with a callback protocol, which also has an unsolved type variable.

**Comment by Drino** [comment]
Thank you again for the fast answer :)

I'm really sorry for bothering you. Let me try to be as concrete as possible this time. 

>You're replacing the Callable with a callback protocol, which also has an unsolved type variable.

It looks like this type variable is resolved in any other case, except the one when I use lambda. 

I'm assuming that lambdas and normal (def) functions are supposed to be more or less the same, but this snippet demonstrates that they are pretty different if lambda is inlined:

```python
from typing import TypeVar, Any, Callable, Protocol
from typing_extensions import reveal_type


TCallable = TypeVar('TCallable', bound=Callable[..., Any])


class Decorator(Protocol):
    def __call__(self, _: TCallable, /) -> TCallable:
        ...


def _decorator_fn(x):
    return x

decorator_fn: Decorator = _decorator_fn

decorator_lambda: Decorator = lambda x: x


@decorator_fn
def decorated_with_fn(x: int) -> str:
    return str(x)


# Argument of type "(x: int) -> str" cannot be assigned to parameter "x" of type "TCallable@__call__"
#  Type "(x: int) -> str" cannot be assigned to type "TCallable@__call__"
@decorator_lambda
def decorated_with_lambda(x: int) -> str:
    return str(x)


reveal_type(decorated_with_fn)  # Type of "decorated_with_fn" is "(x: int) -> str"
reveal_type(decorated_with_lambda)  # Type of "decorated_with_lambda" is "TCallable@__call__"
```

Shouldn't that be considered a bug?

**Comment by Drino** [comment]
>Here again, v is annotated with a type that has an unsolved type variable T. Do you agree this is a problem?

When I opened this ticket I wasn't expecting it to be a problem. Now I'm confused. 

Before your wrote:
>There are two way that type variables can be replaced by a concrete type:
> ...
>Through a call expression where the type checker "solves" the type variable

Isn't `Callable` (or its higher-level parent) supposed to be the special generic type that solves its type variables every time it's called? 

Consider the two following snippets:
```python
from typing import TypeVar, Callable
from typing_extensions import reveal_type


T = TypeVar("T")

TDecorator = Callable[[T], T]

def decorator_factory() -> TDecorator: ...

reveal_type(decorator_factory()(42))  # Type of "decorator_factory()(42)" is "Unknown"
```
and
```python
from typing import TypeVar, Protocol
from typing_extensions import reveal_type


T = TypeVar("T")

class TDecorator(Protocol):
    def __call__(self, _: T, /) -> T: ...


def decorator_factory() -> TDecorator: ...

reveal_type(decorator_factory()(42))  # Type of "decorator_factory()(42)" is "int"
```

So, are `TDecorator = Callable[[T], T]` and `class TDecorator(Protocol): def __call__(self, _: T, \) -> T: ...` really supposed to be different? In [PEP-544](https://peps.python.org/pep-0544/#callback-protocols) it is said that `Callback protocols and Callable[...] types can be used interchangeably.`.

**Comment by erictraut** [comment]
The statement `TDecorator = Callable[[T], T]` defines a generic type alias named `TDecorator` that has a single type parameter. To use this type alias, you need to supply a type argument, as in `TDecorator[int]` or `TDecorator[T]`. If you do not supply a type argument, a type checker will assume that you meant `TDecorator[Any]`. Pyright internally distinguishes between an explicit `Any` and an implicit `Any` and calls the latter `Unknown`.

The statement `def decorator_factory() -> TDecorator: ...` is effectively declaring `def decorator_factory() -> Callable[..., Any]: ...`, which almost certainly not what you intend. I think what you intend here is `def decorator_factory() -> TDecorator[T]: ...`.

The type alias declaration `TDecorator = Callable[[T], T]` is equivalent to the protocol class declaration:

```python
class TDecorator(Protocol[T]):
    def __call__(self, _: T, /) -> T: ...
```

Note that I've added a class-level type parameter `T`, whereas you omitted this above. As with the type alias, when you use this protocol class, you need to supply a type argument. If you don't, a type checker will assume `Any`.

There is an ambiguity in the Python type system relating to generic `Callable`'s that has received much discussion but no resolution. It relates to function signatures of the form `def func() -> Callable[[T], T]: ...`. The ambiguity is whether `T` is meant to be scoped to `func` or to the `Callable` that `func` returns. If `T` were to appear in the input parameter signature of `func` (as in `def func(v: list[T]) -> Callable[[T], T]: ...`), it would be clear that `T` is scoped to `func`. But if `T` doesn't appear anywhere within the input parameter signature of `func`, most type checkers (including pyright) assume that the intent is to associate `T` with the returned `Callable` rather than `func` itself. This behavior isn't standardized or documented anywhere, and it's based on heuristics that differ from one type checker to another.

**Comment by erictraut** [comment]
You are not alone when it comes to confusion about generic type aliases and type variable scoping. Many Python developers get tripped up on these concepts. I'm working on [PEP 695](https://peps.python.org/pep-0695/) which will hopefully make things much clearer.

**Comment by Drino** [comment]
Wow, I do like the syntactic sugar for generics! :)

It also seems to eliminate the issue for Python 3.12+

But just in case I've created a feature request. Hope it's much easier to understand than this bug:
https://github.com/microsoft/pyright/issues/3803
